HeuristicLab.Encodings.SymbolicExpressionTreeEncoding/3.4/Grammars/SymbolicExpressionGrammarBase.cs

```diff
@@ -281,6 +281,32 @@ protected SymbolicExpressionGrammarBase(string name, string description)
       }
     }
 
+    public virtual void ClearAllowedChildSymbols(ISymbol parent) {
+      bool changed = false;
+      if (allowedChildSymbols.TryGetValue(parent.Name, out var childSymbols)) {
+        changed |= childSymbols.Count > 0;
+        childSymbols.Clear();
+      }
+
+      for (int argumentIndex = 0; argumentIndex < GetMaximumSubtreeCount(parent); argumentIndex++) {
+        var key = Tuple.Create(parent.Name, argumentIndex);
+        if (allowedChildSymbolsPerIndex.TryGetValue(key, out childSymbols)) {
+          changed |= childSymbols.Count > 0;
+          childSymbols.Clear();
+        }
+      }
+
+      if (changed) {
+        ClearCaches();
+        OnChanged();
+      }
+    }
+
+    public virtual void ClearAllAllowedChildSymbols() {
+      foreach(var symbol in Symbols)
+        ClearAllowedChildSymbols(symbol);

```

Comment:
This implementation fires a changed event for each symbol in the grammar, which can be quite costly depending on the action that is triggered from the event.

A better solution would be to query if `allowedChildSymbols` or `allowedChildSymbolsPerIndex` contains any items. If they are empty do nothing, otherwise clear both dictionaries and the caches and fire a changed event. With this solution only one event that indicates the change will be fired.

Reply:
Done